commons-math/src/main/java/org/apache/commons/math4/random/SobolSequenceGenerator.java

```diff
@@ -0,0 +1,326 @@
+     * @throws MathParseException if the content in the stream could not be parsed successfully
+     * @throws IOException if an error occurs while reading from the input stream
+     */
+    public SobolSequenceGenerator(final int dimension, final InputStream is)
+            throws NotStrictlyPositiveException, MathParseException, IOException {
+
+        if (dimension < 1) {
+            throw new NotStrictlyPositiveException(dimension);
+        }
+
+        this.dimension = dimension;
+
+        // init data structures
+        direction = new long[dimension][BITS + 1];
+        x = new long[dimension];
+
+        // initialize the other dimensions with direction numbers from the stream
+        int lastDimension = initFromStream(is);
+        if (lastDimension < dimension) {
+            throw new OutOfRangeException(dimension, 1, lastDimension);
+        }
+    }
+
+    /**
+     * Load the direction vector for each dimension from the given stream.
+     * <p>
+     * The input stream <i>must</i> be an ASCII text containing one
+     * valid direction vector per line.
+     *
+     * @param is the input stream to read the direction vector from
+     * @return the last dimension that has been read from the input stream
+     * @throws IOException if the stream could not be read
+     * @throws MathParseException if the content could not be parsed successfully
+     */
+    private int initFromStream(final InputStream is) throws MathParseException, IOException {
+
+        // special case: dimension 1 -> use unit initialization
+        for (int i = 1; i <= BITS; i++) {
+            direction[0][i] = 1l << (BITS - i);
+        }
+
+        final Charset charset = Charset.forName(FILE_CHARSET);
+        final BufferedReader reader = new BufferedReader(new InputStreamReader(is, charset));
+        int dim = -1;
+
+        try {
+            // ignore first line
+            reader.readLine();
+
+            int lineNumber = 2;
+            int index = 1;
+            String line = null;
+            while ( (line = reader.readLine()) != null) {
+                StringTokenizer st = new StringTokenizer(line, " ");
+                try {
+                    dim = Integer.parseInt(st.nextToken());
+                    if (dim >= 2 && dim <= dimension) { // we have found the right dimension
+                        final int s = Integer.parseInt(st.nextToken());
+                        final int a = Integer.parseInt(st.nextToken());
+                        final int[] m = new int[s + 1];
+                        for (int i = 1; i <= s; i++) {
+                            m[i] = Integer.parseInt(st.nextToken());
+                        }
+                        initDirectionVector(index++, a, m);
+                    }
+
+                    if (dim > dimension) {
+                        return dim;
+                    }
+                } catch (NoSuchElementException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)
Similar issue at line numbers 224.